applications/tasks/src/modules/kt_users.erl

```diff
@@ -0,0 +1,556 @@
+    ,{<<"music_on_hold">>, kz_json:new()}
+    ,{<<"ringtones">>, kz_json:new()}
+    ].
+
+apps(Props) -> maybe_nest(<<"apps">>, Props).
+user_portal(Props) -> maybe_nest(<<"user_portal">>, Props).
+
+-spec maybe_nest(kz_term:ne_binary(), kz_term:proplist()) -> kz_term:proplist().
+maybe_nest(_, []) -> [];
+maybe_nest(Feature, Props) -> [{Feature, kz_json:from_list(Props)}].
+
+%%------------------------------------------------------------------------------
+%% @doc Build Users private doc fields from CSV args
+%% @end
+%%------------------------------------------------------------------------------
+-spec add_private_fields(kt_task:args(), kz_doc:object(), kz_type:ne_binary()) -> kz_doc:object().
+add_private_fields(AccountId ,Doc ,Args) ->
+    AccountDb = kz_util:format_account_id(AccountId, 'encoded'),
+    {MD5, SHA1} = pass_hashes(maps:get(<<"username">>, Args), maps:get(<<"password">>, Args)),
+    AlphaNumName = normalize_alphanum_name(maps:get(<<"first_name">>, Args), maps:get(<<"last_name">>, Args)),
+    PvtOptions = [{'type', kzd_user:type()}
+                 ,{'account_id', AccountId}
+                 ,{'account_db', AccountDb}
+                 ,{'crossbar_doc_vsn', <<"1">>}
+                 ],
+    UpdatedDoc = kz_doc:update_pvt_parameters(Doc, AccountDb, PvtOptions),
+    OtherPvtValues = [{<<"pvt_md5_auth">>, MD5}
+                     ,{<<"pvt_sha1_auth">>, SHA1}
+                     ,{<<"pvt_alphanum_name">>, AlphaNumName}
+                     ],
+    kz_json:set_values(OtherPvtValues, UpdatedDoc).
+
+
+%%------------------------------------------------------------------------------
+%% @doc Validate the CSV args are valid to create a user
+%% @end
+%%------------------------------------------------------------------------------
+-spec validate_user(kz_type:ne_binary(), kz_tasks:args()) -> 'ok' | {'error', kz_type:ne_binary()}.
+validate_user(AccountId, #{<<"username">> := Username}) ->
+    validate_username(AccountId, Username).
+
+%%------------------------------------------------------------------------------
+%% @doc Validate the CSV supplied username is valid
+%% @end
+%%------------------------------------------------------------------------------
+-spec validate_username(kz_term:ne_binary(), kz_type:ne_binary()) -> 'ok' | {'error', kz_type:ne_binary()}.
+validate_username(AccountId, Username) ->
+    LowerUsername = kz_term:to_lower_binary(Username),
+    case is_username_unique(AccountId, Username) of
+        'true' ->
+            check_username_length(AccountId, LowerUsername);
+        'false' ->
+            {'error', <<"Username is not unique">>}
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc Validate the users username is unique
+%% @end
+%%------------------------------------------------------------------------------
+-spec is_username_unique(kz_term:ne_binary(), kz_term:ne_binary()) -> boolean().
+is_username_unique(AccountId, Username) ->
```

Comment:
One idea is to move the validation done in cb_users to kzd_users and call it from here as well (minus using crossbar's Context). Similar to what kzdb_accounts is doing for creating/deleting accounts.